### PR TITLE
Make CCM* and CCM independent

### LINKED ISSUE #8482: Make CCM* and CCM independent
Currently `PSA_WANT_ALG_CCM` and `PSA_WANT_ALG_CCM_STAR_NO_TAG` are de [facto synonyms](https://github.com/Mbed-TLS/mbedtls/blob/development/include/psa/crypto_adjust_config_synonyms.h#L39) - enabling any of the two also enables this other. (This is for historical reasons as initially CCM* did not have its own macro, and in terms of built-ins, both are implemented in `ccm.c`.)

This is a problem because it prevents us from making builds of PSA with all AEADs enabled but no unauthenticated cipher enabled, since CCM is an AEAD (used with `psa_aead_xxx()` functions) but `CCM_STAR_NO_TAG` is an unauthenticated cipher (used with `psa_cipher_xxx()` functions).

This task is to change that and make those two independent. We'll want a test build with CCM* enabled but not CCM and vice-versa, and we'll need to make sure dependency declarations are correct everywhere.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #8482.

## PR checklist

- [ ] **changelog** done
- [ ] **backport** not required
- [ ] **tests** provided: I added 2 new test components

### CODE REVIEW

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1596]: Let's build everything just to be sure there isn't something weird going on in programs.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1599]: I think we still want other test suites to run, most notably `test_suite_psa_crypto`, so I'd like to see `make test` here. What I meant in chat was that I don't think we need `ssl-opt.sh` or `compat.sh` in addition.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1624]: Again, I'd prefer just `make` here, to check programs as well.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1628]: I think we want to run the test suites as well (`make test`).

**Review by mpg — CHANGES_REQUESTED** [review]
Looks mostly good to me, I'm just requesting a bit more testing. (IMO `make` plus `make test` is pretty much the baseline, there are very few components where we want less than that, and I don't think the new components are one of those exceptions.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:1602]: Future-looking note: I think this component will become redundant with the "cipher light only" component (or whatever it's called) when we do #8483 and we'll be able to remove it at that point. Let's keep it until then :)

**Comment by mpg** [comment]
> **changelog** I suspect that this is required, right @mpg ?

Yep!

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:1628]: I skipped the testing part here intentionally because I thought that the final `grep` check was enough to ensure that builtin cipher support was not present. But it's not bad to test everything after all, so I'll add the check here as well ;)

**Comment by valeriosetti** [comment]
Note for reviewers: I know that https://github.com/Mbed-TLS/mbedtls/pull/8489/commits/953bee666f170572cc443ba0d2aeb09512f1c0f7 is not correctly handling the acceleration for CCM*, but there is a specific issue for that (https://github.com/Mbed-TLS/mbedtls/issues/8488) so I thought it could be nicer to fix all of them there. Wdyt?

**Comment by mpg** [comment]
> Note for reviewers: I know that [953bee6](https://github.com/Mbed-TLS/mbedtls/commit/953bee666f170572cc443ba0d2aeb09512f1c0f7) is not correctly handling the acceleration for CCM*, but there is a specific issue for that (#8488) so I thought it could be nicer to fix all of them there. Wdyt?

I think it's OK to make the minimum amount of changes needed to make the CI pass here, and leave the rest for #8488.

**Comment by valeriosetti** [comment]
I cannot check the InternalCI, but since OpenCI fully passed I think that the API-ABI break reported by the InternalCI is spurious. So this PR is ready for review

File: library/ssl_misc.h
Comment (mpg) [inline library/ssl_misc.h:297]: Pre-exising: while at it, can you remove `DES_C` from the list? I should have been removed in 3.0 but apparently we missed it. (I checked, and I think that's the only occurrence. I guess we originally missed it because we grepped `library/ssl*.c` instead of `library/ssl*` - I just did the same mistake when checking for other occurrences and first didn't understand why it was not finding this one...)

File: tests/include/test/psa_exercise_key.h
Comment (mpg) [inline tests/include/test/psa_exercise_key.h:55]: Pre-existing: it looks like we're missing an `#else` before this `#undef`.

File: ChangeLog.d/8482.txt
Comment (mpg) [inline ChangeLog.d/8482.txt:1]: I'm not sure about this. The behaviour doesn't change in the default config, which has both enabled anyway. I'm not sure what the best category is, I'm inclined to go for "Changes" since it's officially defined as a catch-all.

**Review by mpg — CHANGES_REQUESTED** [review]
I mostly have minor remarks about the ChangeLog and pre-existing issues I noticed while reviewing, that I suggest fixing here as they're both very small even if only mildly related.

The only big thing is once #8495 is merged (it's currently queued) we'll have conflicts about `SSL_HAVE` macros. We'll need to make a choice about the scheme for these macros. I suggest keeping the scheme introduced in #8495 but I'm happy to hear arguments to the contrary :)

File: library/ssl_misc.h
Comment (valeriosetti) [reply library/ssl_misc.h:281]: Yes, I realized that there were going to be conflicts while reviewing #8495 earlier. I agree that the scheme proposed there is better than mine, so I will align to that one. Perhaps I can just wait for it to be merged and then rebase on `development`. Meanwhile I will fix other comments ;)

File: ChangeLog.d/8482.txt
Comment (valeriosetti) [reply ChangeLog.d/8482.txt:1]: I wasn't sure either, but to me it looked like a behavior change because even thought they were both defined in the default config, they are no more "connected to each other" in the sense that one does not imply the other.
Anyway if you prefer `Changes` that's fine for me ;)

**Comment by valeriosetti** [comment]
I haven't addressed [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8489#discussion_r1387681562) yet, but I'll do that as soon as #8495 is merged ;)

File: ChangeLog.d/8482.txt
Comment (gilles-peskine-arm) [reply ChangeLog.d/8482.txt]: “Default behavior changes” is for incompatible changes in documented behavior. It's similar to “API changes”, but we use “API changes” when affected programs typically won't compile anymore, vs “Default behavior changes” when affected programs will still compile but behave differently.

Here there's no change in documented behavior, so no changelog entry is needed. However, it's plausible that people might have relied on the undocumented behavior that enabling CCM auto-enabled CCM\*, so it's ok to have a notice in the changelog to say this won't happen anymore. That would belong under “Changes” since it doesn't really belong anywhere else.

File: library/ssl_misc.h
Comment (valeriosetti) [reply library/ssl_misc.h:274]: Done now ;)

**Comment by valeriosetti** [comment]
`Windows-2013` failed in the OpenCI, but it seems to be related to some CI issue, so I would say the PR is ready for another round of reviews.

**Review by mpg — DISMISSED** [review]
LGTM, thanks for addressing my comments including the two pre-existing issues!

**Comment by valeriosetti** [comment]
I just rebased to solve conflicts with `development`

**Review by mpg — APPROVED** [review]
LGTM

**Comment by valeriosetti** [comment]
@gabor-mezei-arm the only error on the OpenCI is related to the CI itself, not to the code, so the PR is ready for review ;)

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

File: tests/include/test/psa_exercise_key.h
Comment (mpg) [reply tests/include/test/psa_exercise_key.h:1]: Good catch! However, it's pre-existing and unrelated to this PR, so I'd rather not hold delay this PR for that, considering there are other issues which depend on this.

**Pull request merged**
